File: 3186-maximum-total-damage-with-spell-casting/3186-maximum-total-damage-with-spell-casting.cpp
class Solution {
public:
    long long maximumTotalDamage(vector<int>& power) {
        if (power.empty()) return 0LL;

        sort(power.begin(), power.end());
        vector<int> vals;
        vector<long long> earn;

        for (int i = 0; i < (int)power.size(); ) {
            int v = power[i];
            long long s = 0;
            while (i < (int)power.size() && power[i] == v) {
                s += v;
                ++i;
            }
            vals.push_back(v);
            earn.push_back(s);
        }

        int n = (int)vals.size();
        vector<long long> dp(n + 1, 0);

        for (int i = n - 1; i >= 0; --i) {
            int nextIdx = int(lower_bound(vals.begin(), vals.end(), vals[i] + 3) - vals.begin());
            long long take = earn[i] + dp[nextIdx];
            long long skip = dp[i + 1];
            dp[i] = max(skip, take);
        }
        return dp[0];
    }
};

File: 3186-maximum-total-damage-with-spell-casting/3186-maximum-total-damage-with-spell-casting.py
class Solution:
    def maximumTotalDamage(self, power):
        if not power:
            return 0

        power.sort()
        vals, earn = [], []

        i = 0
        while i < len(power):
            v = power[i]
            total = 0
            while i < len(power) and power[i] == v:
                total += v
                i += 1
            vals.append(v)
            earn.append(total)

        n = len(vals)
        dp = [0] * (n + 1)

        for i in range(n - 1, -1, -1):
            next_idx = bisect_left(vals, vals[i] + 3)
            take = earn[i] + dp[next_idx]
            skip = dp[i + 1]
            dp[i] = max(take, skip)

        return dp[0]